src/krb5/_set_password.pyi

```diff
@@ -1,31 +1,112 @@
+    """
+
+    properties: ADPolicyInfoProp
+    min_length: int
+    history: int
+    max_age: int
+    min_age: int
+
+    @classmethod
+    def from_bytes(cls, data: bytes) -> "ADPolicyInfo":
+        """Decode AD policy result from byte string
+
+        Args:
+            data: Serialized AD policy `server_response`
+
+        Returns:
+            ADPolicyInfo: Decoded AD policy result strcture
+
+        Raises:
+            ValueError: Invalid data length or signature not 0x0000
+        """
+    @classmethod
+    def to_bytes(cls, policy: "ADPolicyInfo") -> bytes:
+        """Reverses the `from_bytes` operation
+
+        Args:
+            policy: AD policy result structure
+
+        Returns:
+            bytes: Serialized AD policy result byte string
+        """
+
+class SetPasswordResultCode(enum.IntEnum):
+    SUCCESS = 0
+    MALFORMED = 1
+    HARDERROR = 2
+    AUTHERROR = 3
+    SOFTERROR = 4
+
 class SetPasswordResult(typing.NamedTuple):
     """The result returned by :meth:`set_password()` and
     :meth:`set_password_using_ccache()`.
 
-    The `result_code` and `result_code_string` is the library response:\n
-    KRB5_KPASSWD_SUCCESS   (0) - Success\n
-    KRB5_KPASSWD_MALFORMED (1) - Malformed request error\n
-    KRB5_KPASSWD_HARDERROR (2) - Server error\n
-    KRB5_KPASSWD_AUTHERROR (3) - Authentication error\n
-    KRB5_KPASSWD_SOFTERROR (4) - Password change rejected\n
+    The `result_code` and `result_code_string` are the pure library responses:
+    - SUCCESS   (0) - Success
+    - MALFORMED (1) - Malformed request error
+    - HARDERROR (2) - Server error
+    - AUTHERROR (3) - Authentication error
+    - SOFTERROR (4) - Password change rejected
 
-    The `result_string` is a server protocol response that may contain useful
+    The `server_response` is a server protocol message that may contain useful
     information about password policy violations or other errors.
+    Depending on `kpasswd` implementation, it may be returned as:\n
+    - decoded UTF-8 string (MIT KDC)
+    - decoded `ADPolicyInfo` (Active Directory Policy Information)
+    - raw bytes (if unable to decode)
+
+    When (and only when) the server response is exactly 30 bytes long starting
+    with `0x0000`, it is assumed to be `ADPolicyInfo`.
+    All other cases are first decoded as UTF-8 string and even if this fails,
+    the raw bytes are returned as `server_response`.
+
+    See `ADPolicyInfo` for more information.
     """
 
-    result_code: int
+    result_code: SetPasswordResultCode
     """The library result code of the password change operation."""
-    result_code_string: bytes
-    """The byte string representation of the result code."""
-    result_string: bytes
-    """Server response string"""
+    result_code_string: str | bytes
+    """The decoded or byte string representation of the result code."""
+    server_response: str | ADPolicyInfo | bytes
```

Comment:
I'm reluctant to set a polymorphic return value here. I would probably err on the side of keeping the raw response from the actual API call as is and having the caller deal with the various types it could be. They are also more involved in their environment and have a better idea as to what might be returned rather than us trying to deal with all possibilities.

Reply:
done